lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+const Ajv = require('ajv');
+const _ = require('lodash');
+const schema = require('../configSchema');
+const errorMessageBuilder = require('./ConfigSchemaHandler/errorMessageBuilder');
+
+const functionNamePattern = '^[a-zA-Z0-9-_]+$';
+
+class ConfigSchemaHandler {
+  constructor(serverless) {
+    this.serverless = serverless;
+    this.schema = _.cloneDeep(schema);
+
+    deepFreeze(this.schema.properties.service);
+    deepFreeze(this.schema.properties.plugins);
+    deepFreeze(this.schema.properties.package);
+    Object.freeze(this.schema.properties.layers);
+    Object.freeze(this.schema.properties.resources);
+  }
+
+  validateConfig(userConfig) {
+    const ajv = new Ajv();
+    const validate = ajv.compile(this.schema);
+    validate(userConfig);
+    if (validate.errors) {
+      const messages = errorMessageBuilder.buildErrorMessages(validate.errors);
+      for (const message of messages) {
+        switch (this.serverless.service.configValidationMode) {
+          case 'error':
+            throw new this.serverless.classes.Error(message);
+          case 'warn':
+            this.serverless.cli.log(message, 'Serverless', { color: 'red' });
+            break;
+          default:
+            throw new this.serverless.classes.Error(
+              '.configValidationMode should be equal to one of the allowed values `error` or `warn`'
```

Comment:
Makes sense. Adjusted it in https://github.com/serverless/serverless/pull/7335/commits/d70a131be0f988d8940c85a055ceac78e721e330

Reply:
No, I had to rollback those changes.  It needs some default `warn` value otherwise an error is shown to user in case `configValidationMode` is not specified (which is the most common scenario).

Anyway, in case there _is_ a way to make it more elegant, it can be done in future PRs